internal/runtime/client/client.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package client provides the Runtime SDK client.
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/client-go/transport"
+
+	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1alpha1"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
+)
+
+const defaultDiscoveryTimeout = 10 * time.Second
+
+// Options are creation options for a Client.
+type Options struct {
+	Catalog *catalog.Catalog
+}
+
+// New returns a new Client.
+func New(options Options) Client {
+	return &client{
+		catalog: options.Catalog,
+	}
+}
+
+// Client is the runtime client to interact with hooks and extensions.
+type Client interface {
+	// Discover makes the discovery call on the extension and updates the runtime extensions
+	// information in the extension status.
+	// TODO: Need a final decision on if we also want to run register inside discover.
+	Discover(context.Context, *runtimev1.ExtensionConfig) (*runtimev1.ExtensionConfig, error)
+}
+
+var _ Client = &client{}
+
+type client struct {
+	catalog *catalog.Catalog
+}
+
+func (c *client) Discover(ctx context.Context, ext *runtimev1.ExtensionConfig) (*runtimev1.ExtensionConfig, error) {
+	gvh, err := c.catalog.GroupVersionHook(runtimehooksv1.Discovery)
+	if err != nil {
+		return nil, err
+	}
+
+	request := &runtimehooksv1.DiscoveryHookRequest{}
+	response := &runtimehooksv1.DiscoveryHookResponse{}
+
+	// Future work: The discovery runtime extension could be operating on a different hook version than
+	// the latest. We will have to loop through different versions of the discover hook here to actually
+	// finish discovery.
+	opts := &httpCallOptions{
+		catalog: c.catalog,
+		config:  ext.Spec.ClientConfig,
+		gvh:     gvh,
+		timeout: defaultDiscoveryTimeout,
+	}
+
+	if err := httpCall(ctx, request, response, opts); err != nil {
+		return nil, errors.Wrap(err, "failed to call the discovery extension")
+	}
+
+	modifiedExtension := &runtimev1.ExtensionConfig{}
+	ext.DeepCopyInto(modifiedExtension)
+	modifiedExtension.Status.Handlers = []runtimev1.ExtensionHandler{}
+	for _, extension := range response.Extensions {
```

Comment:
```suggestion
	for _, handler := range response.Extensions {
```
(makes more sense once we adjusted the DiscoveryReponse struct as suggested above)